stingray/pulse/pulsar.py

```diff
@@ -303,7 +304,107 @@ def profile_stat(profile, err=None):
     return np.sum((profile - mean) ** 2 / err ** 2)
 
 
-def z_n(phase, n=2, norm=1):
+@functools.lru_cache(maxsize=128)
+def _cached_sin_harmonics(nbin, z_n_n):
+    dph = 1.0 / nbin
+    twopiphases = np.pi * 2 * np.arange(dph / 2, 1, dph)
+    cached_sin = np.zeros(z_n_n * nbin)
+    for i in range(z_n_n):
+        cached_sin[i * nbin: (i + 1) * nbin] = np.sin(twopiphases)
+    return cached_sin
+
+
+@functools.lru_cache(maxsize=128)
+def _cached_cos_harmonics(nbin, z_n_n):
+    dph = 1.0 / nbin
+    twopiphases = np.pi * 2 * np.arange(dph / 2, 1, dph)
+    cached_cos = np.zeros(z_n_n * nbin)
+    for i in range(z_n_n):
+        cached_cos[i * nbin: (i + 1) * nbin] = np.cos(twopiphases)
+    return cached_cos
+
+
+@jit(nopython=True)
+def _z_n_fast_cached_sums_unnorm(prof, ks, cached_sin, cached_cos):
+    all_zs = np.zeros(ks.size)
+    N = prof.size
+
+    total_sum = 0
+    for k in ks:
+        local_z = (
+            np.sum(cached_cos[: N * k: k] * prof) ** 2
+            + np.sum(cached_sin[: N * k: k] * prof) ** 2
+        )
+        total_sum += local_z
+        all_zs[k - 1] = total_sum
+
+    return all_zs
+
+
+def z_n_poisson_all(profile, nmax=20):
```

Comment:
So is this the binned Poisson version? I'm trying to figure out the distinction between "poisson" and "events"? If it is the binned events version, I'd probably call it "binned_events" rather than "poisson", just to make clear the distinction with the unbinned version. :)

Reply:
Done